UltraStar Play/Assets/Common/Audio/Recording/CamdAudioSamplesAnalyzer.cs

```diff
@@ -0,0 +1,171 @@
+using System;
+using System.Collections.Generic;
+using UniRx;
+using UnityEngine;
+
+public class CamdAudioSamplesAnalyzer : IAudioSamplesAnalyzer
+{
+    /** There are 49 halftones in the hearable audio spectrum (C2 to C6 (1046.5023 Hz)). */
+    private const int NumHalftones = 49;
+    /** A4 concert pitch of 440 Hz. */
+    private const float BaseToneFreq = 440f;
+    private const int BaseToneMidi = 33;
+    private const int MinSampleLength = 256;
+    private static readonly double[] halftoneFrequencies = PrecalculateHalftoneFrequencies();
+
+    private readonly int[] halftoneDelays;
+    private readonly Subject<PitchEvent> pitchEventStream;
+    private readonly List<int> pitchRecordHistory = new List<int>();
+    private readonly int pitchRecordHistoryLength = 5;
+
+    private bool isEnabled;
+    private int lastPitchDetectedFrame;
+
+    public CamdAudioSamplesAnalyzer(Subject<PitchEvent> pitchEventStream, int sampleRateHz)
+    {
+        this.pitchEventStream = pitchEventStream;
+        halftoneDelays = PrecalculateHalftoneDelays(halftoneFrequencies, sampleRateHz);
+    }
+
+    private static double[] PrecalculateHalftoneFrequencies()
+    {
+        double[] noteFrequencies = new double[NumHalftones];
+        for (int index = 0; index < NumHalftones; index++)
+        {
+            noteFrequencies[index] = BaseToneFreq * Math.Pow(2f, (index - BaseToneMidi) / 12f);
+        }
+        return noteFrequencies;
+    }
+
+    private static int[] PrecalculateHalftoneDelays(double[] halftoneFrequencies, double sampleRateHz)
+    {
+        int[] noteDelays = new int[NumHalftones];
+        for (int index = 0; index < NumHalftones; index++)
+        {
+            noteDelays[index] = Convert.ToInt32((sampleRateHz) / halftoneFrequencies[index]);
+        }
+        return noteDelays;
+    }
+
+    public void Enable()
+    {
+        isEnabled = true;
+    }
+
+    public void Disable()
+    {
+        isEnabled = false;
+    }
+
+    public void ProcessAudioSamples(float[] audioSamplesBuffer, int samplesSinceLastFrame)
+    {
+        if (!isEnabled || samplesSinceLastFrame < MinSampleLength)
+        {
+            return;
```

Comment:
The samples of this frame are discarded here. I think it would be better to buffer them until the buffer has enough samples to be analyzed.

Otherwise (in theory), if the framerate is so high that the samplesSinceLastFrame are always below the MinSampleLength, then it will never be analyzed.

At the moment the framerate is aimed at 30fps, because videos rarely have a higher frame rate. But this could change in the future.

Reply:
The leftover samples will not be usable for improving pitch detection, and thus can be discarded. Storing these samples in a separate array introduces cpu overhead for partially copying array content, for locking / threaded protected access to some PreviousUnusedSamples array, and either for adding clipping detection or for twice running the camd algorithm (for the "old"/stored part and the new samples).
I'd suggest just discarding the additional samples here, because the quality result will already be good enough here anyways.

What definitely should be done (not part of this pr) is to decouple pitch note averaging from screen frames and instead add all detected pitch values of one usdx file pitch + averaging them, then show that as sung note for that beat duration and grade that averaged pitch comparing to the "perfect" note as mentioned in the song ultrastar txt file. This will greatly improve perceived pitch detection / grading quality especially in noisy environments and increase the amounts of fun 😛 . This is already mentioned as TODO in #12 and (in my opinion) should be done outside of the CamdAudioSamplesAnalyzer.cs class.